Fix bug that changed brightness at each HSV update

tempse opened the pull request:
The HSV setter should accept a percentage for the brightness
value but actually assumed the brightness to be in absolute values
between 1 and 255.
This resulted in brightness reductions at each HSV update, in
steps of 100% -> 100/255=39% -> 39/255=15% -> ... (see also
https://github.com/home-assistant/home-assistant/issues/15582,
where I originally reported this bug).

Commit:
Fix bug that changed brightness at each hsv update

The HSV setter should accept a percentage for the brightness
value but actually assumed the brightness to be in absolute values
between 1 and 255.
This resulted in brightness reductions at each HSV update, in
steps of 100% -> 100/255=39% -> 39/255=15% -> ... (see also
https://github.com/home-assistant/home-assistant/issues/15582,
where I originally reported this bug).

kirichkov commented:
If you've tested and you can confirm it's working I'm OK with it. @rytilahti should merge, bump version and you can probably prepare a pull request to HASS.

tempse commented:
Cool, thanks a lot!
My tests were successful and I found everything to be working as expected. As far as I see it, everything should be ready to merge.

rytilahti commented:
Looking at the `fakes.py` it seems that brightness is indeed provided as a percentage integer, so this change sounds fine to me.

Does the reporting work fine for you? `pyhs100 hsv` or simply `pyhs100 state` works for testing. Looking at smartbulb.py's `hsv` it is also doing the conversion. I'm wondering if that's intentionally done, and there are different versions of bulbs working differently, making this a breaking change?

tempse commented:
To me, the reporting seems fine (and consistent with the actual behavior of the bulb).

Here an example with the relevant terminal output:

```
# setting the bulb to fully saturated red at 100% brightness:
% pyhs100 --host <bulb IP> --bulb hsv 0 100 100
Setting HSV: 0 100 100

# actual lighting is good and consistent with the report:
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Color temperature: 0
Is dimmable: True
HSV: (0, 100, 255)
Brightness: 100

# switching colors
% pyhs100 --host <bulb IP> --bulb hsv 240 100 100
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Brightness: 100
Is dimmable: True
Color temperature: 0
HSV: (240, 100, 255)

# changing brightness via HSV setting
% pyhs100 --host <bulb IP> --bulb hsv 240 100 50
Setting HSV: 240 100 50
% pyhs100 --host 192.168.1.117 --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Is dimmable: True
Color temperature: 0
Brightness: 50
HSV: (240, 100, 127)

# changing brightness via the brightness command:
% pyhs100 --host <bulb IP> --bulb brightness 100
Setting brightness to 100
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Is dimmable: True
HSV: (240, 100, 255)
Color temperature: 0
Brightness: 100
```

The reported HSV brightness values are in 1..255 integers, which seems to be connected with the conversion in `smartbulb.py`. My modifications in this PR have not changed this behavior. The only thing which has changed is that repeated calls of the HSV setter do not reduce the brightness anymore. The following is an example of HSV setter calls **before** my code changes:

```
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Color temperature: 0
Brightness: 100
HSV: (0, 100, 255)
Is dimmable: True

% pyhs100 --host <bulb IP> --bulb hsv 0 100 100
Setting HSV: 0 100 100
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Brightness: 39
Color temperature: 0
HSV: (0, 100, 99)
Is dimmable: True
```

It seems to me that my changes shouldn't break anything as they do not affect the internal representations of the brightness values. But I do not have other TP-Link bulbs for explicitly testing this claim, unfortunately...
I'll let you decide whether we should merge now or wait for someone to actually test it.

tempse commented:
To me, the reporting seems fine (and consistent with the actual behavior of the bulb).

Here an example with the relevant terminal output:

```
# setting the bulb to fully saturated red at 100% brightness:
% pyhs100 --host <bulb IP> --bulb hsv 0 100 100
Setting HSV: 0 100 100

# actual lighting is good and consistent with the report:
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Color temperature: 0
Is dimmable: True
HSV: (0, 100, 255)
Brightness: 100

# switching colors
% pyhs100 --host <bulb IP> --bulb hsv 240 100 100
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Brightness: 100
Is dimmable: True
Color temperature: 0
HSV: (240, 100, 255)

# changing brightness via HSV setting
% pyhs100 --host <bulb IP> --bulb hsv 240 100 50
Setting HSV: 240 100 50
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Is dimmable: True
Color temperature: 0
Brightness: 50
HSV: (240, 100, 127)

# changing brightness via the brightness command:
% pyhs100 --host <bulb IP> --bulb brightness 100
Setting brightness to 100
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Is dimmable: True
HSV: (240, 100, 255)
Color temperature: 0
Brightness: 100
```

The reported HSV brightness values are in 1..255 integers, which seems to be connected with the conversion in `smartbulb.py`. My modifications in this PR have not changed this behavior. The only thing which has changed is that repeated calls of the HSV setter do not reduce the brightness anymore. The following is an example of HSV setter calls **before** my code changes:

```
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Color temperature: 0
Brightness: 100
HSV: (0, 100, 255)
Is dimmable: True

% pyhs100 --host <bulb IP> --bulb hsv 0 100 100
Setting HSV: 0 100 100
% pyhs100 --host <bulb IP> --bulb state
== SmartBulb - LB130(EU) ==
Device state: ON
Host/IP: <bulb IP>
Brightness: 39
Color temperature: 0
HSV: (0, 100, 99)
Is dimmable: True
```

It seems to me that my changes shouldn't break anything as they do not affect the internal representations of the brightness values. But I do not have other TP-Link bulbs for explicitly testing this claim, unfortunately...
I'll let you decide whether we should merge now or wait for someone to actually test it.

tempse commented:
Subtly trying to revive this issue, is there any chance of merging this PR and propagating everything to HASS, @rytilahti?

(TLDR of my last message: I am pretty sure that this bugfix cannot affect other devices as it does not change the underlying choice of the units of brightness.)

rytilahti commented:
So my only worry is that this changes the behaviour, currently both output and input are given in 0-255 (meaning that only the apidoc and the test client is wrong on this matter). So the simplest solution would be to adapt home-assistant to follow this format.

However, I think that percentage-based API is better (and consistent with brightness property), so I'd be for merging this after changing the `hsv` property to return the value also in percentages.

Btw, considering you have a test device around, would you have interest in getting #98 and #119 merged?

Commit:
Modify HSV property to return brightness in percent

Switch from reported brightness values of 1..255 to percentage
values, for consistency with the apidoc and 8761dd8.

tempse commented:
> So my only worry is that this changes the behaviour, currently both output and input are given in 0-255 (meaning that only the apidoc and the test client is wrong on this matter). So the simplest solution would be to adapt home-assistant to follow this format.

Yes, the behavior changes. However, to the best of my judgement, I corrected a bug in my first commit of this PR which is actually disconnected from this discussion. In fact, before my initial changes, it was impossible to set the bulb to full brightness via the HSV setter method. (The highest possible value to enter was 100, which was internally converted to 100/255=39%. Input values >100 were considered invalid and triggered an error.)

> However, I think that percentage-based API is better (and consistent with brightness property), so I'd be for merging this after changing the hsv property to return the value also in percentages.

You are right, if this PR is merged, the HSV getter should consistently report brightness in percent. Please see my last commit for the corresponding changes.

> Btw, considering you have a test device around, would you have interest in getting #98 and #119 merged?

Sure, I am happy to take a look as soon as I can find a little more time to spare. :+1:

rytilahti commented:
> The highest possible value to enter was 100, which was internally converted to 100/255=39%. Input values >100 were considered invalid and triggered an error.

Hmm, where was the error triggered (besides the cli tool)? If I'm reading the code correctly, inputing 255 as a parameter would have yielded 100%.

> You are right, if this PR is merged, the HSV getter should consistently report brightness in percent.

That's great, and I'm ready to merge this. However, could you also include the received json payload for the hsv query to `fakes.py`, disallow having values outside 0-100 range and add a test for that?

> Sure, I am happy to take a look as soon as I can find a little more time to spare.

Great! Thanks for your contribution! :-)

tempse commented:
> Hmm, where was the error triggered (besides the cli tool)? If I'm reading the code correctly, inputing 255 as a parameter would have yielded 100%.

Actually, it was [impossible](https://github.com/GadgetReactor/pyHS100/blob/master/pyHS100/cli.py#L143) to enter a brightness value of 255. The following output is produced by the package before my initial changes (repo state: v0.3.2, e612560b):
```
% pyhs100 --ip <IP> --bulb hsv 0 100 255
Usage: pyhs100 hsv [OPTIONS] H S V

Error: Invalid value for "v": 255 is not in the valid range of 0 to 100.
```
Apart from the CLI, this error also shows up in other applications (like Home Assistant, where I initially reported this).

> That's great, and I'm ready to merge this. However, could you also include the received json payload for the hsv query to fakes.py, disallow having values outside 0-100 range and add a test for that?

I checked all files related to brightness (i.e., `cli.py`, `smartbulb.py`, `smartplug.py`, `tests/fakes.py`, `test_bulb.py`) and found brightness values and ranges to be consistently defined in percentage values in all of them.

To wrap it up, the only thing that seemed off were the few unit conversions in `smartbulb.py`, which I removed in this PR. Since everything else has always been defined as percentages, am I correct in saying that no further changes are required for merging?

rytilahti commented:
> Apart from the CLI, this error also shows up in other applications (like Home Assistant, where I initially reported this).

Yes, the cli tool was enforcing the value, but I think it is not used by as many people as the library itself.

> I checked all files related to brightness (i.e., cli.py, smartbulb.py, smartplug.py, tests/fakes.py, test_bulb.py) and found brightness values and ranges to be consistently defined in percentage values in all of them.

Ok, great!

> To wrap it up, the only thing that seemed off were the few unit conversions in smartbulb.py, which I removed in this PR. Since everything else has always been defined as percentages, am I correct in saying that no further changes are required for merging?

That's correct, could you please add guards to enforce that the input values are as expected (and raise exceptions accordingly when not, `SmartDeviceException` class with a informational message would be nice), and add `test_hsv` for testing the valid and invalid values (`pyHS100/tests/test_bulb.py`)? I will then merge and create a new release with this change :-)

aaronasmith commented:
Any chance this will be merged soon?  I have one of these lights and the constant dimming is getting old.

rytilahti commented:
There are those small outstanding issues I mentioned in my last comment that needs to be taken care of before getting this merged. ping @tempse (in case the comment got buried :-))

Commit:
Add checks and tests for the hsv setter

- make sure that new (hue, saturation, brightness) values are
  within their valid ranges (0..255, 0..100, 0..100) and raise
  SmartDeviceException if they are not
- add test function for the hsv setter

tempse commented:
I added checks for the input value ranges of the HSV setter as well as unit tests for this function. `SmartDeviceException` is raised when invalid input is passed.

If the range checks are not well placed in the code, I might need some help with where to best put them. Let me know what you think.

(Also, I'm sorry that this took me so long. I was swamped with work recently…)

rytilahti commented:
Thanks, looks great! :1st_place_medal:  Let's merge this, I'll hope to make a release during this week, so we can get this also updated to downstream :-)

tempse commented:
Great! And thanks a lot for releasing v0.3.3 so quickly!
